# Preparação de estados

## 1 qubit
Um estado qualquer de um qubit pode ser escrito como
\begin{align}
|\psi\rangle & = c_{0}|0\rangle + c_{1}|1\rangle \\
& = |c_{0}|e^{i\phi_{0}}|0\rangle + |c_{1}|e^{i\phi_{1}}|1\rangle \\
& = e^{i\phi_{0}}\big(|c_{0}||0\rangle + |c_{1}|e^{i(\phi_{1}-\phi_{0})}|1\rangle\big) \\
& = e^{i\phi_{0}}\big(\cos(\theta/2)|0\rangle + \sin(\theta/2)e^{i\phi}|1\rangle\big) \hspace{2cm} (\star) \\
& = e^{i\phi_{0}}e^{i\phi/2}\big(e^{-i\phi/2}\cos(\theta/2)|0\rangle + e^{-i\phi/2}e^{i\phi}\sin(\theta/2)|1\rangle\big) \\
& = e^{i\phi_{0}}e^{i\phi/2}\big(e^{-i\phi/2}\cos(\theta/2)|0\rangle + e^{i\phi/2}\sin(\theta/2)|1\rangle\big) \\
& = e^{i(\phi_{1}+\phi_{0})/2}\big(e^{-i\phi/2}\cos(\theta/2)|0\rangle + e^{i\phi/2}\sin(\theta/2)|1\rangle\big) \\
& = e^{it/2}\big(e^{-i\phi/2}\cos(\theta/2)|0\rangle + e^{i\phi/2}\sin(\theta/2)|1\rangle\big),
\end{align}
em que definimos
\begin{align}
& \phi:=\phi_{1}-\phi_{0}, \\
& t := \phi_{1}+\phi_{0}
\end{align}
e
$$\cos(\theta/2):=|c_{0}| \text{ e } \sin(\theta/2)=|c_{1}|,$$
pois $0\le|c_{j}|\le1$ para $j=0,1$ e $|c_{0}|^{2}+|c_{1}|^{2}=1$. Note que essa definição implica na restrição
$$\theta\in[0,\pi].$$ 

A menos de uma fase global, o estado ($\star$) é determinado por 2 ângulos em coordenadas esféricas na esfera de Bloch. Fazendo uma rotação $R_{z}(-\phi)$ zeramos a coordenada $y$ do vetor de Bloch. Na sequência, aplicando uma rotação $R_{y}(-\theta)$ zeramos a componente $x$ do vetor de Bloch, o levando assim para o eixo $z$, ou seja, obtendo assim o estado $|0\rangle$.

De forma geral, uma rotação por um ângulo $\theta$ em torno da direção $\hat{n}$ é implementada usando
$$R_{\hat{n}}(\theta) = e^{-i\theta \hat{n}\cdot\vec{\sigma}/2} = \cos(\theta/2)\mathbb{I} - i\sin(\theta/2)\hat{n}\cdot\vec{\sigma}.$$
Teremos assim os casos particulares
\begin{align}
& R_{z}(\theta) = \cos(\theta/2)\mathbb{I} - i\sin(\theta/2)\sigma_{z} = \begin{bmatrix} e^{-i\theta/2} & 0 \\ 0 & e^{i\theta/2} \end{bmatrix}, \\
& R_{y}(\theta) = \cos(\theta/2)\mathbb{I} - i\sin(\theta/2)\sigma_{y} = \begin{bmatrix} \cos(\theta/2) & -\sin(\theta/2) \\ \sin(\theta/2) & \cos(\theta/2) \end{bmatrix}.
\end{align}
Ou seja
\begin{align}
& R_{z}(\theta)|0\rangle = e^{-i\theta/2}|0\rangle, \\
& R_{z}(\theta)|1\rangle = e^{i\theta/2}|1\rangle, \\
& R_{y}(\theta)|0\rangle = \cos(\theta/2)|0\rangle+\sin(\theta/2)|1\rangle, \\
& R_{y}(\theta)|1\rangle = -\sin(\theta/2)|0\rangle+\cos(\theta/2)|1\rangle.
\end{align}

Com isso, vem que
\begin{align}
R_{z}(-\phi)|\psi\rangle & = e^{it/2}\big(e^{-i\phi/2}\cos(\theta/2)R_{z}(-\phi)|0\rangle + e^{i\phi/2}\sin(\theta/2)R_{z}(-\phi)|1\rangle\big) \\
& = e^{it/2}\big(e^{-i\phi/2}\cos(\theta/2)e^{i\phi/2}|0\rangle + e^{i\phi/2}\sin(\theta/2)e^{-i\phi/2}|1\rangle\big) \\
& = e^{it/2}\big(\cos(\theta/2)|0\rangle + \sin(\theta/2)|1\rangle\big), \\
R_{y}(-\theta)R_{z}(-\phi)|\psi\rangle & = e^{it/2}\big(\cos(\theta/2)R_{y}(-\theta)|0\rangle + \sin(\theta/2)R_{y}(-\theta)|1\rangle\big) \\
& = e^{it/2}\big(\cos(\theta/2)\big(\cos(\theta/2)|0\rangle-\sin(\theta/2)|1\rangle\big) + \sin(\theta/2)\big(\sin(\theta/2)|0\rangle+\cos(\theta/2)|1\rangle\big)\big) \\
& = e^{it/2}|0\rangle.
\end{align}
Por conseguinte
\begin{align}
& |0\rangle = e^{-it/2}R_{y}(-\theta)R_{z}(-\phi)|\psi\rangle, \\
& |\psi\rangle = e^{it/2}R_{z}(\phi)R_{y}(\theta)|0\rangle =  R_{z}(\phi)R_{y}(\theta)XP(t/2)X|0\rangle,
\end{align}
com 
\begin{align}
& \theta=2\arccos(|c_{0}|),\\
& \phi=\phi_{1}-\phi_{0}, \\
& t=\phi_{1}+\phi_{0}, \\
& \phi_{j} = \arctan(\Im(c_{j})/\Re(c_{j})),\ j=0,1.
\end{align}
Se $\Re(c_{j})=0$ então $\phi_j=\pi/2$.

Acima usamos a porta de fase
$$P(\phi)=|0\rangle\langle 0|+e^{i\phi}|1\rangle\langle 1|.$$

Vale observar que se os coeficientes são reais, i.e.,
$$|\psi\rangle = r_{0}|0\rangle+r_{1}|1\rangle,$$
então podemos prepará-lo usando
$$|\psi\rangle = R_{y}(\theta)|0\rangle$$
com
$$\theta = 2\arccos(r_0).$$

__Exercício:__ Fazer o algoritmo e implementar a preparação de um estado puro qualquer de 1 qubit, usando o método descrito acima.

## 2 qubits
Um estado puro geral de 2 qubits pode ser escrito como segue
\begin{align}
|\psi\rangle & = c_{00}|00\rangle + c_{01}|01\rangle + c_{10}|10\rangle + c_{11}|11\rangle \\
& = |0\rangle\big(c_{00}|0\rangle + c_{01}|1\rangle\big) + |1\rangle\big(c_{10}|0\rangle + c_{11}|1\rangle\big) \\
& = r_{0}|0\rangle r_{0}^{-1}\big(c_{00}|0\rangle + c_{01}|1\rangle\big) + r_{1}|1\rangle r_{1}^{-1}\big(c_{10}|0\rangle + c_{11}|1\rangle\big) \\
& =: r_{0}|0\rangle|\psi_{0}\rangle + r_{1}|1\rangle|\psi_1\rangle,
\end{align}
com
\begin{align}
& r_{0}^{2} := |c_{00}|^{2} + |c_{01}|^{2}, \\
& r_{1}^{2} := |c_{10}|^{2} + |c_{11}|^{2}.
\end{align}
Assim como fizemos para 1 qubit, aqui podemos escrever
\begin{align}
|\psi_0\rangle & = r_{0}^{-1}\big(c_{00}|0\rangle + c_{01}|1\rangle\big) \\
& = r_{0}^{-1}\big(|c_{00}|e^{i\phi_{00}}|0\rangle + |c_{01}|e^{i\phi_{01}}|1\rangle\big) \\
& = r_{0}^{-1}e^{it_{0}/2}\big(|c_{00}|e^{-i\phi_{0}/2}|0\rangle + |c_{01}|e^{i\phi_{0}/2}|1\rangle\big),
\end{align}
com
\begin{align}
& t_0 := \phi_{01}+\phi_{00}, \\ 
& \phi_{0}  := \phi_{01}-\phi_{00}.
\end{align}
Teremos assim que
\begin{align}
R_{z}(-\phi_0)|\psi_0\rangle & = r_{0}^{-1}e^{it_{0}/2}\big(|c_{00}|e^{-i\phi_{0}/2}R_{z}(-\phi_0)|0\rangle + |c_{01}|e^{i\phi_{0}/2}R_{z}(-\phi_0)|1\rangle\big) \\
& = r_{0}^{-1}e^{it_{0}/2}\big(|c_{00}|e^{-i\phi_{0}/2}e^{i\phi_{0}/2}|0\rangle + |c_{01}|e^{i\phi_{0}/2}e^{-i\phi_{0}/2}|1\rangle\big) \\
& = r_{0}^{-1}e^{it_{0}/2}\big(|c_{00}||0\rangle + |c_{01}||1\rangle\big).
\end{align}

Não obstante, devemos notar que neste caso
$$|c_{00}|^{2} + |c_{01}|^{2} =: r_{0}^{2}\le 1.$$
Vamos escrever então
$$|c_{00}| = r_{0}\cos(\theta_0/2) \text{ e } |c_{01}| = r_{0}\sin(\theta_0/2).$$
Com isso, vem que
\begin{align}
R_{z}(-\phi_0)|\psi_0\rangle & = r_{0}^{-1}r_{0}e^{it_{0}/2}\big(\cos(\theta_0/2)|0\rangle + \sin(\theta_0/2)|1\rangle\big) \\
& = e^{it_{0}/2}\big(\cos(\theta_0/2)|0\rangle + \sin(\theta_0/2)|1\rangle\big).
\end{align}
Assim
\begin{align}
& R_{y}(-\theta_0)R_{z}(-\phi_0)|\psi_0\rangle \\
& = e^{it_{0}/2}\big(\cos(\theta_0/2)R_{y}(-\theta_0)|0\rangle + \sin(\theta_0/2)R_{y}(-\theta_0)|1\rangle\big) \\
& = e^{it_{0}/2}\big(\cos(\theta_0/2)\big(\cos(\theta_0/2)|0\rangle-\sin(\theta_0/2)|1\rangle\big) + \sin(\theta_0/2)\big(\sin(\theta_0/2)|0\rangle+\cos(\theta_0/2)|1\rangle\big)\big) \\
& = e^{it_{0}/2}|0\rangle.
\end{align}
Portanto, se $r_{0}\ne 0$, teremos
\begin{align}
|0\rangle & = e^{-it_{0}/2}R_{y}(-\theta_0)R_{z}(-\phi_0)|\psi_0\rangle, \\
|\psi_0\rangle & = e^{it_{0}/2}R_{z}(\phi_0)R_{y}(\theta_0)|0\rangle \\
& =: U_{0}|0\rangle,
\end{align}
com
\begin{align}
& t_{0}=\phi_{01}+\phi_{00}, \\
& \phi_{0}=\phi_{01}-\phi_{00}, \\
& \theta_0 = 2\arctan(|c_{01}|/|c_{00}|).
\end{align}

Analogamente, pode-se mostrar que
\begin{align}
|0\rangle & = e^{-it_{1}/2}R_{y}(-\theta_1)R_{z}(-\phi_1)|\psi_1\rangle, \\
|\psi_1\rangle & = r_{1}^{-1}\big(c_{10}|0\rangle + c_{11}|1\rangle\big) \\
& = e^{it_{1}/2}R_{z}(\phi_1)R_{y}(\theta_1)|0\rangle \\
& =: U_{1}|0\rangle,
\end{align}
com
\begin{align}
& t_{1}=\phi_{11}+\phi_{10}, \\
& \phi_{1}=\phi_{11}-\phi_{10}, \\
& \theta_1 = 2\arctan(|c_{11}|/|c_{10}|).
\end{align}

Notemos que
$$r_{0}^{2}+r_{1}^{2} = |c_{00}|^{2} + |c_{01}|^{2} + |c_{10}|^{2} + |c_{11}|^{2} =1.$$
Como 
$$0\le r_0,r_1\le 1$$
podemos fazer
$$r_0 =: \cos(\xi/2) \text{ e } r_1 =: \sin(\xi/2).$$
Como
\begin{align}
|\psi\rangle & =  r_{0}|0\rangle|\psi_{0}\rangle + r_{1}|1\rangle|\psi_1\rangle \\
& = r_{0}|0\rangle\otimes U_{0}|0\rangle + r_{1}|1\rangle\otimes U_{1}|0\rangle,
\end{align}

podemos preparar qualquer estado de 2 qubits usando operações unitárias controladas como segue
$$|\psi\rangle = C_{U_0}^{0_0\rightarrow 1}C_{U_1}^{0_1\rightarrow 1}\big(R_y(\xi)\otimes\mathbb{I}\big)\big(|0\rangle\otimes|0\rangle\big)$$
com
\begin{align}
& U_0 = e^{it_{0}/2}R_{z}(\phi_0)R_{y}(\theta_0), \\
& U_1 = e^{it_{1}/2}R_{z}(\phi_1)R_{y}(\theta_1).
\end{align}

Vale observar que para __coeficientes reais__, podemos usar
\begin{align}
& U_0 = R_{y}(\theta_0), \\
& U_1 = R_{y}(\theta_1).
\end{align}
Vale observar também que para preparar estados gerais de 2 qubits, utilizaremos o algoritmo de preparação de estados de 1 qubit com coeficientes reais, para preparar
$$r_{0}|0\rangle + r_{1}|1\rangle = \cos(\xi/2)|0\rangle +\sin(\xi/2)|1\rangle = R_{y}(\xi)|0\rangle.$$

__Exercício:__ Fazer o algoritmo e implementar a preparação de um estado puro qualquer de 2 qubits, usando o método descrito acima.

## 3 qubits

Começamos olhando para
\begin{align}
|\psi\rangle & = c_{000}|000\rangle + c_{001}|001\rangle + c_{010}|010\rangle + c_{011}|011\rangle + c_{100}|100\rangle + c_{101}|101\rangle + c_{110}|110\rangle + c_{111}|111\rangle \\
& = |00\rangle\big(c_{000}|0\rangle + c_{001}|1\rangle\big) + |01\rangle\big(c_{010}|0\rangle + c_{011}|1\rangle\big) + |10\rangle\big(c_{100}|0\rangle + c_{101}|1\rangle\big) + |11\rangle\big(c_{110}|0\rangle + c_{111}|1\rangle\big) \\
& = r_{00}|00\rangle r_{00}^{-1}\big(c_{000}|0\rangle + c_{001}|1\rangle\big) + r_{01}|01\rangle r_{00}^{-1}\big(c_{010}|0\rangle + c_{011}|1\rangle\big) + r_{10}|10\rangle r_{10}^{-1}\big(c_{100}|0\rangle + c_{101}|1\rangle\big) + r_{11}|11\rangle r_{11}^{-1}\big(c_{110}|0\rangle + c_{111}|1\rangle\big) \\
& = r_{00}|00\rangle|\psi_{00}\rangle + r_{01}|01\rangle|\psi_{01}\rangle + r_{10}|10\rangle|\psi_{10}\rangle + r_{11}|11\rangle|\psi_{11}\rangle \\
& = r_{00}|00\rangle\otimes U_{00}|0\rangle + r_{01}|01\rangle\otimes U_{01}|0\rangle + r_{10}|10\rangle\otimes U_{10}|0\rangle + r_{11}|11\rangle\otimes U_{11}|0\rangle \\
\end{align}
em que definimos
\begin{align}
& |\psi_{00}\rangle := r_{00}^{-1}\big(c_{000}|0\rangle + c_{001}|1\rangle\big), \\
& |\psi_{01}\rangle := r_{01}^{-1}\big(c_{010}|0\rangle + c_{011}|1\rangle\big), \\
& |\psi_{10}\rangle := r_{10}^{-1}\big(c_{100}|0\rangle + c_{101}|1\rangle\big), \\
& |\psi_{11}\rangle := r_{11}^{-1}\big(c_{110}|0\rangle + c_{111}|1\rangle\big),
\end{align}
com
\begin{align}
& r_{00}^{2} := |c_{000}|^{2} + |c_{001}|^{2}, \\
& r_{01}^{2} := |c_{010}|^{2} + |c_{011}|^{2}, \\
& r_{10}^{2} := |c_{100}|^{2} + |c_{101}|^{2}, \\
& r_{11}^{2} := |c_{110}|^{2} + |c_{111}|^{2}.
\end{align}

Começamos então usando o algoritmo para preparação de estados de 2 qubits como coeficientes reais para preparar
$$r_{00}|00\rangle + r_{01}|01\rangle + r_{10}|10\rangle + r_{11}|11\rangle =: |\Phi\rangle.$$
Feito isso, podemos preparar o estado geral de 3 qubits usando a seguinte sequência de operações unitárias controladas
$$|\psi\rangle = C_{U_{00}}^{0_{0}1_{0}\rightarrow 2}C_{U_{01}}^{0_{0}1_{1}\rightarrow 2}C_{U_{10}}^{0_{1}1_{0}\rightarrow 2}C_{U_{11}}^{0_{1}1_{1}\rightarrow 2}\big(|\Phi\rangle_{01}\otimes|0\rangle_2\big).$$
Analogamente ao que obtivemos para os casos de 1 e 2 qubits, aqui no caso de 3 qubits teremos
\begin{align}
& U_{00} = e^{it_{00}/2}R_{z}(\phi_{00})R_{y}(\theta_{00}), \\
& U_{01} = e^{it_{01}/2}R_{z}(\phi_{01})R_{y}(\theta_{01}), \\
& U_{10} = e^{it_{10}/2}R_{z}(\phi_{10})R_{y}(\theta_{10}), \\
& U_{11} = e^{it_{11}/2}R_{z}(\phi_{11})R_{y}(\theta_{11}),
\end{align}
com
\begin{align}
& t_{jk} = \phi_{jk0}+\phi_{jk1},\\
& \phi_{jk} = \phi_{jk1}-\phi_{jk0}, \\
& \theta_{jk} = 2\arctan(|c_{jk1}|/|c_{jk0}|)
\end{align}
para $j,k=0,1$, e escrevemos
$$c_{jkl}=|c_{jkl}|e^{i\phi_{jkl}}.$$

__Exercício:__ Fazer o algoritmo e implementar a preparação de um estado puro qualquer de 3 qubits, usando o método descrito acima.

## 4 qubits



In [7]:
list_bin = []
for j in range(0,2**4):
    b = "{:04b}".format(j)
    list_bin.append(b)
print(list_bin)

['0000', '0001', '0010', '0011', '0100', '0101', '0110', '0111', '1000', '1001', '1010', '1011', '1100', '1101', '1110', '1111']


In [8]:
list_int = []
for j in range(0,2**4):
    list_int.append(int(list_bin[j],2))
print(list_int)

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15]


Teremos assim
\begin{align}
|\psi\rangle & = c_{0000}|0000\rangle + c_{0001}|0001\rangle + c_{0010}|0010\rangle + c_{0011}|0011\rangle + c_{0100}|0100\rangle + c_{0101}|0101\rangle + c_{0110}|0110\rangle + c_{0111}|0111\rangle \\
& + c_{1000}|1000\rangle + c_{1001}|1001\rangle + c_{1010}|1010\rangle + c_{1011}|1011\rangle + c_{1100}|1100\rangle + c_{1101}|1101\rangle + c_{1110}|1110\rangle + c_{1111}|1111\rangle \\
& = |000\rangle\big(c_{0000}|0\rangle + c_{0001}|1\rangle\big) + |001\rangle\big(c_{0010}|0\rangle + c_{0011}|1\rangle\big) + |010\rangle\big(c_{0100}|0\rangle + c_{0101}|1\rangle\big) + |011\rangle\big(c_{0110}|0\rangle + c_{0111}|1\rangle\big) \\
& + |100\rangle\big(c_{1000}|0\rangle + c_{1001}|1\rangle\big) + |101\rangle\big(c_{1010}|0\rangle + c_{1011}|1\rangle\big) + |110\rangle\big(c_{1100}|0\rangle + c_{1101}|1\rangle\big) + |111\rangle\big(c_{1110}|0\rangle + c_{1111}|1\rangle\big) \\
& = r_{000}|000\rangle r_{000}^{-1}\big(c_{0000}|0\rangle + c_{0001}|1\rangle\big) + r_{001}|001\rangle r_{001}^{-1}\big(c_{0010}|0\rangle + c_{0011}|1\rangle\big) + r_{010}|010\rangle r_{010}^{-1}\big(c_{0100}|0\rangle + c_{0101}|1\rangle\big) \\
& + r_{011}|011\rangle r_{011}^{-1}\big(c_{0110}|0\rangle + c_{0111}|1\rangle\big) + r_{100}|100\rangle r_{100}^{-1}\big(c_{1000}|0\rangle + c_{1001}|1\rangle\big) + r_{101}|101\rangle r_{101}^{-1}\big(c_{1010}|0\rangle \\
& + c_{1011}|1\rangle\big) + r_{110}|110\rangle r_{110}^{-1}\big(c_{1100}|0\rangle + c_{1101}|1\rangle\big) + r_{111}|111\rangle r_{111}^{-1}\big(c_{1110}|0\rangle + c_{1111}|1\rangle\big) \\
& = r_{000}|000\rangle|\psi_{000}\rangle + r_{001}|001\rangle|\psi_{001}\rangle  + r_{010}|010\rangle|\psi_{010}\rangle + r_{011}|011\rangle|\psi_{011}\rangle \\
& + r_{100}|100\rangle|\psi_{100}\rangle + r_{101}|101\rangle|\psi_{101}\rangle + r_{110}|110\|\psi_{110}\rangle + r_{111}|111\rangle|\psi_{111}\rangle \\
& = r_{000}|000\rangle U_{000}|0\rangle + r_{001}|001\rangle U_{001}|0\rangle  + r_{010}|010\rangle U_{010}|0\rangle + r_{011}|011\rangle U_{011}|0\rangle \\
& + r_{100}|100\rangle U_{100}|0\rangle + r_{101}|101\rangle U_{101}|0\rangle + r_{110}|110\rangle U_{110}|0\rangle + r_{111}|111\rangle U_{111}|0\rangle \\
& = C_{U_{000}}^{0_{0}1_{0}2_0\rightarrow 3}C_{U_{001}}^{0_{0}1_{0}2_1\rightarrow 3}C_{U_{010}}^{0_{0}1_{1}2_0\rightarrow 3}C_{U_{011}}^{0_{0}1_{1}2_1\rightarrow 3}C_{U_{100}}^{0_{1}1_{0}2_0\rightarrow 3}C_{U_{101}}^{0_{1}1_{0}2_1\rightarrow 3}C_{U_{110}}^{0_{1}1_{1}2_0\rightarrow 3}C_{U_{111}}^{0_{1}1_{1}2_1\rightarrow 3}\big(|\Phi\rangle_{012}\otimes|0\rangle_3\big)
\end{align}

com
\begin{align}
& |\psi_{000}\rangle = r_{000}^{-1}\big(c_{0000}|0\rangle + c_{0001}|1\rangle\big), \\
& |\psi_{001}\rangle = r_{001}^{-1}\big(c_{0010}|0\rangle + c_{0011}|1\rangle\big), \\
& |\psi_{010}\rangle = r_{010}^{-1}\big(c_{0100}|0\rangle + c_{0101}|1\rangle\big), \\
& |\psi_{011}\rangle = r_{011}^{-1}\big(c_{0110}|0\rangle + c_{0111}|1\rangle\big), \\
& |\psi_{100}\rangle = r_{100}^{-1}\big(c_{1000}|0\rangle + c_{1001}|1\rangle\big), \\
& |\psi_{101}\rangle = r_{101}^{-1}\big(c_{1010}|0\rangle + c_{1011}|1\rangle\big) , \\
& |\psi_{110}\rangle = r_{110}^{-1}\big(c_{1100}|0\rangle + c_{1101}|1\rangle\big), \\
& |\psi_{111}\rangle = r_{111}^{-1}\big(c_{1110}|0\rangle + c_{1111}|1\rangle\big).
\end{align}
e 
\begin{align}
|\Phi\rangle & = r_{000}|000\rangle + r_{001}|001\rangle  + r_{010}|010\rangle + r_{011}|011\rangle \\
& + r_{100}|100\rangle + r_{101}|101\rangle + r_{110}|110\rangle + r_{111}|111\rangle.
\end{align}

Ademais, temos que 
$$U_{jkl} = e^{it_{jkl}/2}R_{z}(\phi_{jkl})R_{y}(\theta_{jkl}) $$
com
\begin{align}
& t_{jkl} = \phi_{jkl0}+\phi_{jkl1},\\
& \phi_{jkl} = \phi_{jkl1}-\phi_{jkl0}, \\
& \theta_{jkl} = 2\arctan(|c_{jkl1}|/|c_{jkl0}|)
\end{align}
para $j,k,l=0,1$, e escrevemos
$$c_{jklm}=|c_{jklm}|e^{i\phi_{jklm}}.$$

## $n$ qubits
Podemos estender os resultados particulares acima para obter um algoritmo geral para preparação de estados puros quaisquer de $n$ qubits:
1. Prepara um estado de $n-1$ qubits
$$|\psi\rangle_{n-1} = \sum_{j_0,j_1,\cdots,j_{n-2}=0}^{1}r_{j_0,j_1,\cdots,j_{n-2}}|j_0,j_1,\cdots,j_{n-2}\rangle$$
com coeficientes reais 
$$r_{j_0,j_1,\cdots,j_{n-2}} = \sqrt{|c_{j_0,j_1,\cdots,j_{n-2},0}|^{2} + |c_{j_0,j_1,\cdots,j_{n-2},1}|^{2}}.$$

2. Aplica $2^{n-1}$ portas multicontroladas com os primeiros $n-1$ qubits como controle e com o último qubit como target:
$$C_{U_{j_0,j_1,\cdots,j_{n-2}}}^{0_{j_0},1_{j_1},\cdots,(n-2)_{j_{n-2}}\rightarrow n-1}$$
para 
$$j_0,j_1,\cdots,j_{n-2}=0,1$$
e
$$U_{j_0,j_1,\cdots,j_{n-2}} = e^{it_{j_0,j_1,\cdots,j_{n-2}}/2}R_{z}(\phi_{j_0,j_1,\cdots,j_{n-2}})R_{y}(\theta_{j_0,j_1,\cdots,j_{n-2}})$$
em que
\begin{align}
& t_{j_0,j_1,\cdots,j_{n-2}} = \phi_{j_0,j_1,\cdots,j_{n-2},1} + \phi_{j_0,j_1,\cdots,j_{n-2},0}, \\
& \phi_{j_0,j_1,\cdots,j_{n-2}} = \phi_{j_0,j_1,\cdots,j_{n-2},1} - \phi_{j_0,j_1,\cdots,j_{n-2},0}, \\
& \theta_{j_0,j_1,\cdots,j_{n-2}} = 2\arctan\big(|c_{j_0,j_1,\cdots,j_{n-2},1}|/|c_{j_0,j_1,\cdots,j_{n-2},0}|\big), \\
\end{align}

Vale notar que para a preparação de estados com coeficientes reais, usamos o mesmo algoritmo mas com
$$U_{j_0,j_1,\cdots,j_{n-2}} = R_{y}(\theta_{j_0,j_1,\cdots,j_{n-2}})$$
com
$$\theta_{j_0,j_1,\cdots,j_{n-2}} = 2\arctan\big(c_{j_0,j_1,\cdots,j_{n-2},1}/c_{j_0,j_1,\cdots,j_{n-2},0}\big).$$

Maneiras otimizadas que implementar as operações unitárias envolvidas neste algoritmo estão discutidas na ref. 2.

## La pregunta
Tem uma forma de fazer a preparação de estados sem recorrer a estas operações multicontroladas?

### Referências

1. https://qiskit.org/documentation/tutorials/circuits/3_summary_of_quantum_operations.html

1. V. V. Shende, S. S. Bullock, and I. L. Markov, “Synthesis of Quantum Logic Circuits,” IEEE Trans. Comput.-Aided Des. Integr. Circuits Syst., vol. 25, no. 6, pp. 1000–1010, Jun. 2006, doi: 10.1109/TCAD.2005.855930.